zebra-chain/src/transaction/txidhash.rs

```diff
@@ -0,0 +1,75 @@
+use std::io;
+
+use super::Transaction;
+
+use crate::serialization::{sha256d, ZcashSerialize};
+
+use super::Hash;
+
+pub(super) struct TxIdHasher<'a> {
+    trans: &'a Transaction,
+}
+
+impl<'a> TxIdHasher<'a> {
+    pub fn new(trans: &'a Transaction) -> Self {
+        TxIdHasher { trans }
+    }
+
+    pub(super) fn txid(self) -> Result<Hash, io::Error> {
+        match self.trans {
+            Transaction::V1 { .. }
+            | Transaction::V2 { .. }
+            | Transaction::V3 { .. }
+            | Transaction::V4 { .. } => self.txid_v4(),
+            Transaction::V5 { .. } => self.txid_v5(),
+        }
+    }
+
+    fn txid_v4(self) -> Result<Hash, io::Error> {
```

Comment:
Can we make the pre-v5 name a bit more accurate?

Also let's pass the correct branch id for v5 as a precaution against future use.
```suggestion
            | Transaction::V4 { .. } => self.txid_v1_to_v4(),
            Transaction::V5 { network_upgrade, .. } => self.txid_v5(network_upgrade),
        }
    }

    fn txid_v1_to_v4(self) -> Result<Hash, io::Error> {
```